README
NAME
    Crypt::SSLeay - OpenSSL support for LWP

SYNOPSIS
        lwp-request https://www.example.com

        use LWP::UserAgent;
        my $ua  = LWP::UserAgent->new;
        my $response = $ua->get('https://www.example.com/');
        print $response->content, "\n";

DESCRIPTION
    This document describes "Crypt::SSLeay" version 0.57_05, released
    2010-08-15.

    This Perl module provides support for the HTTPS protocol under LWP, to
    allow an "LWP::UserAgent" object to perform GET, HEAD and POST requests.
    Please see LWP for more information on POST requests.

    The "Crypt::SSLeay" package provides "Net::SSL", which is loaded by
    "LWP::Protocol::https" for https requests and provides the necessary SSL
    glue.

    This distribution also makes following deprecated modules available:

        Crypt::SSLeay::CTX
        Crypt::SSLeay::Conn
        Crypt::SSLeay::X509

    Work on Crypt::SSLeay has been continued only to provide https support
    for the LWP (libwww-perl) libraries.

ENVIRONMENT VARIABLES
    The following environment variables change the way "Crypt::SSLeay" and
    "Net::SSL" behave.

        # proxy support
        $ENV{HTTPS_PROXY} = 'http://proxy_hostname_or_ip:port';

        # proxy_basic_auth
        $ENV{HTTPS_PROXY_USERNAME} = 'username';
        $ENV{HTTPS_PROXY_PASSWORD} = 'password';

        # debugging (SSL diagnostics)
        $ENV{HTTPS_DEBUG} = 1;

        # default ssl version
        $ENV{HTTPS_VERSION} = '3';

        # client certificate support
        $ENV{HTTPS_CERT_FILE} = 'certs/notacacert.pem';
        $ENV{HTTPS_KEY_FILE}  = 'certs/notacakeynopass.pem';

        # CA cert peer verification
        $ENV{HTTPS_CA_FILE}   = 'certs/ca-bundle.crt';
        $ENV{HTTPS_CA_DIR}    = 'certs/';

        # Client PKCS12 cert support
        $ENV{HTTPS_PKCS12_FILE}     = 'certs/pkcs12.pkcs12';
        $ENV{HTTPS_PKCS12_PASSWORD} = 'PKCS12_PASSWORD';

INSTALL
  OpenSSL
    You must have OpenSSL or SSLeay installed before compiling this module.
    You can get the latest OpenSSL package from <http://www.openssl.org/>.

    On Debian systems, you will need to install the libssl-dev package, at
    least for the duration of the build (it may be removed afterwards).

    Other package-based systems may require something similar. The key is
    that "Crypt::SSLeay" makes calls to the OpenSSL library, and how to do
    so is specified in the C header files that come with the library. Some
    systems break out the header files into a separate package from that of
    the libraries. Once the program has been built, you don't need the
    headers any more.

    When installing openssl make sure your config looks like:

        ./config --openssldir=/usr/local/openssl

    or

        ./config --openssldir=/usr/local/ssl

    If you are planning on upgrading the default OpenSSL libraries on a
    system like RedHat, (not recommended), then try something like:

        ./config --openssldir=/usr --shared

    The "--shared" option to config will set up building the .so shared
    libraries which is important for such systems. This is followed by:

        make
        make test
        make install

    This way Crypt::SSLeay will pick up the includes and libraries
    automatically. If your includes end up going into a separate directory
    like /usr/local/include, then you may need to symlink
    /usr/local/openssl/include to /usr/local/include

  Crypt::SSLeay
    The latest Crypt::SSLeay can be found at your nearest CPAN, as well as
    <http://search.cpan.org/dist/Crypt-SSLeay/>

    Once you have downloaded it, Crypt::SSLeay installs easily using the
    "make" * commands as shown below.

        perl Makefile.PL
        make
        make test
        make install

    On Windows systems, both Strawberry Perl and ActiveState (as a separate
    download via ppm) projects include a MingW based compiler distribution
    and "dmake" which can be used to build both OpenSSL and "Crypt-SSLeay".
    If you have such a set up, use "dmake" above.

    For unattended (batch) installations, to be absolutely certain that
    Makefile.PL does not prompt for questions on STDIN, set the following
    environment variable beforehand:

        PERL_MM_USE_DEFAULT=1

    (This is true for any CPAN module that uses "ExtUtils::MakeMaker").

    To skip live tests, you can use

        perl Makefile.PL --no-live-tests

    and to force live tests, you can use

        perl Makefile.PL --live-tests

   Windows
    "Crypt::SSLeay" builds correctly with Strawberry Perl.

    For Activestate users, the ActiveState company does not have a permit
    from the Canadian Federal Government to distribute cryptographic
    software. This prevents "Crypt::SSLeay" from being distributed as a PPM
    package from their repository. See
    <http://aspn.activestate.com/ASPN/docs/ActivePerl/5.8/faq/ActivePerl-faq
    2.html#crypto_packages> for more information on this issue.

    You may download it from Randy Kobes's PPM repository by using the
    following command:

        ppm install http://theoryx5.uwinnipeg.ca/ppms/Crypt-SSLeay.ppd

    An alternative is to add the uwinnipeg.ca PPM repository to your local
    installation. See <http://cpan.uwinnipeg.ca/htdocs/faqs/ppm.html> for
    more details.

   VMS
    It is assumed that the OpenSSL installation is located at "/ssl$root".
    Define this logical to point to the appropriate place in the filesystem.

PROXY SUPPORT
    LWP::UserAgent and Crypt::SSLeay have their own versions of proxy
    support. Please read these sections to see which one is appropriate.

  LWP::UserAgent proxy support
    LWP::UserAgent has its own methods of proxying which may work for you
    and is likely to be incompatible with Crypt::SSLeay proxy support. To
    use LWP::UserAgent proxy support, try something like:

        my $ua = LWP::UserAgent->new;
        $ua->proxy([qw( https http )], "$proxy_ip:$proxy_port");

    At the time of this writing, libwww v5.6 seems to proxy https requests
    fine with an Apache mod_proxy server. It sends a line like:

        GET https://www.example.com HTTP/1.1

    to the proxy server, which is not the CONNECT request that some proxies
    would expect, so this may not work with other proxy servers than
    mod_proxy. The CONNECT method is used by Crypt::SSLeay's internal proxy
    support.

  Crypt::SSLeay proxy support
    For native Crypt::SSLeay proxy support of https requests, you need to
    set the environment variable "HTTPS_PROXY" to your proxy server and
    port, as in:

        # proxy support
        $ENV{HTTPS_PROXY} = 'http://proxy_hostname_or_ip:port';
        $ENV{HTTPS_PROXY} = '127.0.0.1:8080';

    Use of the "HTTPS_PROXY" environment variable in this way is similar to
    "LWP::UserAgent-"env_proxy()> usage, but calling that method will likely
    override or break the Crypt::SSLeay support, so do not mix the two.

    Basic auth credentials to the proxy server can be provided this way:

        # proxy_basic_auth
        $ENV{HTTPS_PROXY_USERNAME} = 'username';
        $ENV{HTTPS_PROXY_PASSWORD} = 'password';

    For an example of LWP scripting with "Crypt::SSLeay" native proxy
    support, please look at the eg/lwp-ssl-test script in the
    "Crypt::SSLeay" distribution.

CLIENT CERTIFICATE SUPPORT
    Client certificates are supported. PEM0encoded certificate and private
    key files may be used like this:

        $ENV{HTTPS_CERT_FILE} = 'certs/notacacert.pem';
        $ENV{HTTPS_KEY_FILE}  = 'certs/notacakeynopass.pem';

    You may test your files with the eg/net-ssl-test program, bundled with
    the distribution, by issuing a command like:

        perl eg/net-ssl-test -cert=certs/notacacert.pem \
            -key=certs/notacakeynopass.pem -d GET $HOST_NAME

    Additionally, if you would like to tell the client where the CA file is,
    you may set these.

        $ENV{HTTPS_CA_FILE} = "some_file";
        $ENV{HTTPS_CA_DIR}  = "some_dir";

    Note that, if specified, $ENV{HTTPS_CA_FILE} must point to the actual
    certificate file. That is, $ENV{HTTPS_CA_DIR} is *not* the path were
    $ENV{HTTPS_CA_FILE} is located.

    For certificates in $ENV{HTTPS_CA_DIR} to be picked up, follow the
    instructions on
    <http://www.openssl.org/docs/ssl/SSL_CTX_load_verify_locations.html>

    There is no sample CA cert file at this time for testing, but you may
    configure eg/net-ssl-test to use your CA cert with the -CAfile option.
    (TODO: then what is the ./certs directory in the distribution?)

  Creating a test certificate
    To create simple test certificates with OpenSSL, you may run the
    following command:

        openssl req -config /usr/local/openssl/openssl.cnf \
            -new -days 365 -newkey rsa:1024 -x509 \
            -keyout notacakey.pem -out notacacert.pem

    To remove the pass phrase from the key file, run:

        openssl rsa -in notacakey.pem -out notacakeynopass.pem

  PKCS12 support
    The directives for enabling use of PKCS12 certificates is:

        $ENV{HTTPS_PKCS12_FILE}     = 'certs/pkcs12.pkcs12';
        $ENV{HTTPS_PKCS12_PASSWORD} = 'PKCS12_PASSWORD';

    Use of this type of certificate takes precedence over previous
    certificate settings described. (TODO: unclear? Meaning "the presence of
    this type of certificate??)

SSL versions
    Crypt::SSLeay tries very hard to connect to *any* SSL web server
    accomodating servers that are buggy, old or simply not
    standards-compliant. To this effect, this module will try SSL
    connections in this order:

    SSL v23
        should allow v2 and v3 servers to pick their best type

    SSL v3
        best connection type

    SSL v2
        old connection type

    Unfortunately, some servers seem not to handle a reconnect to SSL v3
    after a failed connect of SSL v23 is tried, so you may set before using
    LWP or Net::SSL:

        $ENV{HTTPS_VERSION} = 3;

    to force a version 3 SSL connection first. At this time only a version 2
    SSL connection will be tried after this, as the connection attempt order
    remains unchanged by this setting.

ACKNOWLEDGEMENTS
    Many thanks to the following individuals who helped improve
    Crypt-SSLeay:

    Gisle Aas
        for writing this module and many others including libwww, for perl.
        The web will never be the same :)

    Ben Laurie
        deserves kudos for his excellent patches for better error handling,
        SSL information inspection, and random seeding.

    Dongqiang Bai
        for host name resolution fix when using a proxy.

    Stuart Horner of Core Communications, Inc.
        who found the need for building "--shared" OpenSSL libraries.

    Pavel Hlavnicka
        for a patch for freeing memory when using a pkcs12 file, and for
        inspiring more robust "read()" behavior.

    James Woodyatt
        is a champ for finding a ridiculous memory leak that has been the
        bane of many a Crypt::SSLeay user.

    Bryan Hart
        for his patch adding proxy support, and thanks to Tobias Manthey for
        submitting another approach.

    Alex Rhomberg
        for Alpha linux ccc patch.

    Tobias Manthey
        for his patches for client certificate support.

    Daisuke Kuroda
        for adding PKCS12 certificate support.

    Gamid Isayev
        for CA cert support and insights into error messaging.

    Jeff Long
        for working through a tricky CA cert SSLClientVerify issue.

    Chip Turner
        for patch to build under perl 5.8.0.

    Joshua Chamas
        for the time he spent maintaining the module.

    Jeff Lavallee
        for help with alarms on read failures (CPAN bug #12444).

    Guenter Knauf
        for significant improvements in configuring things in Win32 and
        Netware lands and Jan Dubois for various suggestions for
        improvements.

    and others
        who provided bug reports, suggestions, fixes and patches.

SEE ALSO
    Net::SSL
        If you have downloaded this distribution as of a dependency of
        another distribution, it's probably due to this module (which is
        included in this distribution).

    Net::SSLeay
        Net::SSLeay provides access to the OpenSSL API directly from Perl.
        See <http://search.cpan.org/dist/Net-SSLeay/>.

    OpenSSL binary packages for Windows
        See <http://www.openssl.org/related/binaries.html>.

SUPPORT
    For use of Crypt::SSLeay & Net::SSL with Perl's LWP, please send email
    to libwww@perl.org <mailto:libwww@perl.org>.

    For OpenSSL or general SSL support, including issues associated with
    building and installing OpenSSL on your system, please email the OpenSSL
    users mailing list at openssl-users@openssl.org
    <mailto:openssl-users@openssl.org>. See
    <http://www.openssl.org/support/community.html> for other mailing lists
    and archives.

    Please report all bugs at
    "/rt.cpan.org/NoAuth/Bugs.html?Dist=Crypt-SSLeay"" in "http:.

    This module was originally written by Gisle Aas, and was subsequently
    maintained by Joshua Chamas, David Landgren, brian d foy and Sinan Unur.

COPYRIGHT
    Copyright (c) 2010 A. Sinan Unur

    Copyright (c) 2006-2007 David Landgren.

    Copyright (c) 1999-2003 Joshua Chamas.

    Copyright (c) 1998 Gisle Aas.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Crypt-SSLeay
=================================

0.57_04

- Reorganize Makefile.PL to break individual steps in to subroutines so as to
  facilitate localized future modifications. Whether there was any point to
  this remains to be seen.
- In Makefile.PL, accommodate ActiveState+MingW configuration
- Pick the correct lib location for Strawberry Perl (bug #60230).
- Apply fixes in bug reports #59312 and #33954 to Net/SSL.pm.
- Bump version number in Net/SSL.pm to indicate development release.

0.57_03 2010-08-09 20:12:30

- If the last component of $inc_dir is 'openssl', set $inc_dir to its parent
  and set $prefix to 'openssl'. Otherwise, leave $inc_dir as is and set $prefix
  to '' (I hope this fixes bugs #28431, #28680, #32084, #43084, #54103 without
  breaking anything).
- Improved OpenSSL detection on Win32/Strawberry Perl (bug #49285).
- Add MIME::Base64 as a prerequisite as Net::SSL needs it.
- exit 0 if OpenSSL can't be found to avoid superfluous reports from CPAN
  Testers.
- Add \ to $opt_bench in eg/net-ssl-test (bugs #30931, #39363).

0.57_02 2010-08-08 18:27:40 UTC

- Refactor the version detection algorithm in Makefile.PL to handle all known
  variations of version number encoding. Should fix bug #52408. 

0.57_01 2008-02-18 14:42:32 UTC

- use #include <..> rather than #include "..." in
  crypt_ssleay_version.h.
- add command-line switch to avoid live tests (bug #30268).
- skip tests in t/01-connect.t if 443 is already in use (bug #30985).
- make code gcc -Wwrite-strings compatible (bug #31926).

0.57 2007-09-17 20:45:20 UTC

- Honour both $ENV{NO_PROXY} and $ENV{no_proxy} in
  Net::SSL::proxy_connect_helper. (Bug #29371 reported by Jan Dubois).
- $@ construction used in Net::SSL::connect was messed up, which could
  lead to sub-optimal error reporting. (Bug #29372 reported by Jan
  Dubois).
- Ensure no proxification is used in t/01-connect.t (which might be the
  reason for all the spurious smoke failures). Bug #29373 reported by,
  you guessed it, Jan Dubois).
- Silence a dubious fopen() warning in SSLeay.xs.
- s/Netware/NetWare/ in Net/SSL.pm platform check
- Improvements to Makefile.PL for dealing with platforms where openssl
  is installed with ./include and ./lib as sibling directories rather
  than child directories. This should allow the code to configure and
  build "out of the box" on Solaris (and probably other proprietary
  platforms).
- Don't carp in LICENSE key addition code in Makefile.PL

0.56_01 2007-08-09 21:59:47 UTC

- Various improvements to the Win32 configure code in Makefile.PL,
  based on CPAN tickets #28431 and #28432, by Guenter Knauf,
  notably to allow static linking and OpenSSL living in a relative
  directory.
- Net::SSL: alarm() is not implemented on Netware platform, so don't
  try to set one there. (Guenter Knauf).
- Should build on Solaris correctly (bug fix in include dir
  specification). Based on fix suggested in CPAN bug #28680.

0.56 2007-07-10 19:08:20 UTC

- Purely a documentation issue raised by CPAN bug #27935. Users
  of previous versions do not need to upgrade.

0.55 2007-06-01 17:34:22 UTC

- Added a blocking() method to Net::SSL (and bumped version to
  2.81).

0.54 2007-04-12 22:05:26 UTC

- Rebadged 0.53_05, since no bugs appear to have surfaced.

0.53_05

- Fixed up incorrect LIBS key in WriteMakefile args. Thanks to
  David Cantrell for giving me access to an OpenBSD box that
  revealed this problem.
- Added the list of modules that depend on Crypt::SSLeay to
  the README, as per cpants.perl.org. (think: improvements
  to the test suite).

0.53_04 2007-03-06 09:39:01 UTC

- add diag() info to determine possible reasons for failure as per
  http://www.nntp.perl.org/group/perl.cpan.testers/2007/03/msg428964.html
- Tweaks for Strawberry Perl detection.

0.53_03 2007-03-04 18:30:06 UTC

- Adjusted the typemap shims to silence the compiler warnings that
  occur when sizeof(IV) is larger than sizeof(char *).
- use XSLoader for faster loading if available, otherwise fall
  back to DynaLoader.
- Makefile.PL heavily reworked, lots of cruft removed.
- Ask to see whether the live tests should be run.
- renamed net_sst.t to 01-connect.t
- added 02-live.t that performs live HTTPS requests.

0.53_02 2007-01-29 10:02:34 UTC

- don't proxy hosts in NO_PROXY environment variable (CPAN
  bug #11078).
- don't send user agent string to proxy unless
  send_useragent_to_proxy is enabled. (CPAN bug #4759).
- Net::SSL bumped to 2.80

0.53_01 2007-01-24 22:21:09 UTC

- patch for CPAN #12444 applied (Jeff Lavallee). Net::SSL bumped
  tp 2.79.
- example scripts moved into eg/ directory and the documentation
  updated.
- added a TODO to remind me of what needs to be done.

0.53 2006-12-26 17:21:22 UTC

- 0.52_02 deemed stable

0.52_02 2006-12-20 19:29:01 UTC

- improved VMS support (CPAN bug #19829).
- add a test to see if cert file is readable in
  Net::SSL::configure_certs (CPAN bug #8498) and Net::SSL version
  to 2.78.
- known working platforms list removed from documentation. Too old,
  and CPAN Testers has the up-to-date information.
- minor documentation improvements.

0.52_01 2006-12-17

- add call to SSL_library_init() in new()
- maintenance taken over by brian d foy and David Landgren.

$MODULE=Crypt::SSLeay; $VERSION = .51; $DATE="2003-06-10";

- fixed build problem for OpenSSL 0.9.6 and some builds
  of perl 5.8.x which resulted in make error:

   /usr/include/openssl/des.h:193: parse error before '&' token"

  Thanks to Rob Brown for submitting a similar patch to cover this problem

- bug fix from Dongqiang Bai when server using proxy cannot
  resolve host name being connected to

+ Added documentation for updating system OpenSSL libraries 
  for systems such has RedHat that have shared libraries built
  Work sponsored by Stuart Horner of Core Communications, Inc.

$MODULE=Crypt::SSLeay; $VERSION = .49; $DATE="2003-01-30";

+ Documentation updates, including new support address

Makefile.PL
use strict;
use warnings;

use Config;
use ExtUtils::MakeMaker;
use File::Basename qw(basename dirname);
use Getopt::Long;

eval "use ExtUtils::MakeMaker::Coverage";
$@ or print "Adding testcover target\n";

use vars qw($opt_default $opt_libpath $opt_static $opt_live_tests);

GetOptions(
    "default",     \$opt_default,
    "lib=s",       \$opt_libpath,
    "static",      \$opt_static,
    "live-tests!", \$opt_live_tests,
);

$opt_default ||= $ENV{CRYPT_SSLEAY_DEFAULT};

# FIND POSSIBLE SSL INSTALLATIONS

my @POSSIBLE_SSL_DIRS;
if ($opt_libpath) {
    # explicit from command-line
    @POSSIBLE_SSL_DIRS = ($opt_libpath);
    $opt_default       = 1;
}
elsif ($^O eq 'MSWin32') {
    push @POSSIBLE_SSL_DIRS, 'c:\\openssl';
    push @POSSIBLE_SSL_DIRS, $Config{usrinc} if -d $Config{usrinc};
    # pre-Oct2009 Strawberry Perl
    push @POSSIBLE_SSL_DIRS, $Config{prefix};
    push @POSSIBLE_SSL_DIRS, $Config{prefix} . '\..\c';
}
elsif ($^O eq 'VMS') {
    @POSSIBLE_SSL_DIRS = '/ssl$root';
}
else {
    # Unix and the rest
    @POSSIBLE_SSL_DIRS = qw(
        /local
        /local/ssl
        /opt/ssl
        /usr
        /usr/local
        /usr/local/ssl
        /usr/local/openssl
    );
}

my @CANDIDATE = do {
    my %seen;
    grep {!$seen{$_->{check}}++}
        map {Candidate($_)}
            @POSSIBLE_SSL_DIRS
};

if (@CANDIDATE == 0) {
    $opt_default = 0;
    my $bar = '=' x 72;
    print <<"INFO";
$bar
No installed SSL libraries found in any of the following places.
INFO
    print "    $_\n" for @POSSIBLE_SSL_DIRS;
    print <<"INFO";
You will have to either specify a directory location at the following
prompt, or rerun the Makefile.PL program and use the --lib switch
to specify the path. If the path in question is considered standard
on your platform, please consider filing a bug report in order to
have it taken into account in a subsequent version of Crypt::SSLeay.

INFO

    if (-f '/etc/debian_version') {
        print <<"DEBIAN_INFO";

This host looks like it is running Debian. Crypt::SSLeay needs to
be compiled with C headers that the libssl-dev package makes
available. Please install that package before trying to build this
module. (You can always deinstall the package afterwards, once
Crypt::SSLeay has been built).
DEBIAN_INFO
    }
}

my $SSL_DIR;
my $pkg_config;

if (@CANDIDATE == 1) {
    $pkg_config = $CANDIDATE[0];
    if ($opt_default) {
        $SSL_DIR = $pkg_config->{dir};
    }
    else {
        print <<"INFO";
=======================================================
Only one $pkg_config->{type} installation found at $pkg_config->{dir}
Consider running 'perl Makefile.PL --default' the next
time Crypt::SSLeay is upgraded to select this directory
automatically thereby avoiding the following prompt.
=======================================================
INFO
    }
}
elsif (@CANDIDATE > 1) {
    print "Found multiple possibilities for OpenSSL\n";
    for my $c (@CANDIDATE) {
        print "  $c->{dir} ($c->{type} $c->{ver})\n";
    }
}

if (not $SSL_DIR) {
    my %cand;
    for my $c (@CANDIDATE) {
        $cand{$c->{dir}} = {%$c};
    }

    $SSL_DIR = prompt "Which SSL install path do you want to use?",
        $CANDIDATE[0]->{dir};

    if (exists $cand{$SSL_DIR}) {
        # we've already determined that this directory is usable
        $pkg_config = $cand{$SSL_DIR};
    }
    else {
        # unknown directory, better check it out
        $pkg_config = Candidate($SSL_DIR);
        if (not $pkg_config) {
            warn <<"INFO";
$SSL_DIR does not appear to be an SSL library installation, since
the required header files were not found. The build cannot proceed.
INFO
        # avoid failure reports from CPAN Testers
        # when OpenSSL was not found
            exit 0;
        }
    }
}

# note: $SSL_DIR is now sane at this point

my (@INC_FLAGS, @LIB_FLAGS);

if ($^O eq 'VMS') {
    push @INC_FLAGS, "-I$pkg_config->{inc}";
    push @LIB_FLAGS, qw(-L/SYS$SHARE -lSSL$LIBSSL_SHR32 -lSSL$LIBCRYPTO_SHR32);
}
elsif ($^O eq 'MSWin32') {
    # external tools probably expect \ and not / for path separators
    $SSL_DIR =~ tr{/}{\\};

    # default to drive C: if no drive or relative path
    $SSL_DIR = "c:$SSL_DIR" if $SSL_DIR !~ m{\A(?:[a-z]:|\.\.[\\/])}i;

    my $inc = $pkg_config->{inc};
    $inc =~ tr{/}{\\};
    $inc = "c:$inc" if $inc !~ m{\A(?:[a-z]:|\.\.[\\/])}i;
    push @INC_FLAGS, "-I$inc";
    push @INC_FLAGS, "-I$SSL_DIR\\inc32" if -d "$SSL_DIR/inc32";

    my $vanilla = $Config{cc} eq 'gcc' ? 1 : 0;
    $vanilla and print "Assuming Vanilla/Strawberry Perl installation\n";

    if ($vanilla and -d "$SSL_DIR\\lib\\MinGW") {
        push @LIB_FLAGS, "-L$SSL_DIR\\lib\\MinGW";
    }
    elsif(-d "$SSL_DIR/lib") {
        push @LIB_FLAGS, "-L$SSL_DIR\\lib";
    }
    else {
        my $dir = $opt_static ? "$SSL_DIR\\out32" : "$SSL_DIR\\out32dll";
        if (-d $dir) {
            push @LIB_FLAGS, "-L$dir";
        }
        else {
            # Allow developers to point at OpenSSL source...
            push @LIB_FLAGS, "-L$SSL_DIR";
        }
    }

    push @LIB_FLAGS, qw(-lssleay32 -llibeay32);
    push @LIB_FLAGS, qw(-lRSAglue -lrsaref) if $pkg_config->{type} ne 'OpenSSL';
}
else {
    push @INC_FLAGS, "-I$pkg_config->{inc}";

    push @LIB_FLAGS, "-L$pkg_config->{lib}", qw(-lssl -lcrypto -lgcc);
    push @LIB_FLAGS, qw(-lRSAglue -lrsaref) if $pkg_config->{type} ne 'OpenSSL';

    # ccc on alpha support
    if ($^O eq 'linux'
        and `uname -m` =~ /alpha/
        and !(system("nm $SSL_DIR/lib/libssl.a|grep -q 'U _Ots'")>>8)
    ) {
        push @LIB_FLAGS, '-lots';
    }

    # this fix was suggested for building on RedHat 9
    push @INC_FLAGS, '-I/usr/kerberos/include' if -d '/usr/kerberos/include';
}

# write include file that determines ssl support we need to include crypto.h
# for SSLeay so the version gets picked up in SSLeay.xs

open(INCLUDE, ">crypt_ssleay_version.h")
    || die("can't open crypt_ssleay_version.h for writing: $!");
print INCLUDE <<"INFO";
#include <$pkg_config->{prefix}ssl.h>
#include <$pkg_config->{prefix}crypto.h>
#include <$pkg_config->{prefix}err.h>
#include <$pkg_config->{prefix}rand.h>
#include <$pkg_config->{prefix}pkcs12.h>

INFO

if ($] < 5.005) {
    print "adding PL_sv_undef symbol for this ancient perl installation";
    print INCLUDE <<"INFO";
/* defining PL_sv_undef for very old perls ($]) */
#ifndef PL_sv_undef
#define PL_sv_undef sv_undef
#endif

INFO
}

if ($pkg_config->{type} eq 'OpenSSL') {
    # OPENSSL_free defined in OpenSSL 0.9.6 and higher
    if ($pkg_config->{ver} =~ /\b0\.9\.[2-5]/) {
        print INCLUDE "#define CRYPT_SSLEAY_free free\n";
    }
    else {
        print INCLUDE "#define CRYPT_SSLEAY_free OPENSSL_free\n";
    }
}
else {
    print INCLUDE "#define CRYPT_SSLEAY_free free\n";
}

close INCLUDE or die "Cannot close crypt_ssleay_version.h for output: $!\n";

print <<"INFO";

BUILD INFORMATION
================================================
ssl library: $pkg_config->{type} $pkg_config->{ver} in $SSL_DIR
ssl header:  $pkg_config->{prefix}ssl.h
libraries:   @LIB_FLAGS
include dir: @INC_FLAGS
================================================
INFO

my @license =
    do {
        my $version = $ExtUtils::MakeMaker::VERSION;
        $version =~ tr/_//d;
        $version} > 6.30
    ? qw(LICENSE perl)
    : ();

WriteMakefile(
    NAME          => 'Crypt::SSLeay',
    AUTHOR        => 'David Landgren',
    ABSTRACT_FROM => 'SSLeay.pm',
    VERSION_FROM  => 'SSLeay.pm',
    LIBS          => ["@LIB_FLAGS"],
    INC           => "@INC_FLAGS",
    NEEDS_LINKING => 1,
    PREREQ_PM     => {
        'MIME::Base64' => 0, # for Net::SSL
    },
    clean         => {
        FILES => 'crypt_ssleay_version.h test.config',
    },
    @license,
);

if (open OUT, '> test.config') {
    print OUT <<"INFO";
ssl $pkg_config->{type} $pkg_config->{ver} in $SSL_DIR
lib @LIB_FLAGS
inc @INC_FLAGS
cc $Config{cc}
INFO

    my $network_tests = 'n';
    if (not defined $opt_live_tests) {
        print <<"INFO";
The test suite can attempt to connect to public servers
to ensure that the code is working properly. If you are
behind a strict firewall or have no network connectivity,
these tests may fail (through no fault of the code).
INFO

        $network_tests = prompt "Do you want to run the live tests (y/N) ?", 'N';
    }
    elsif ($opt_live_tests) {
        $network_tests = 'y';
    }

    print OUT "network_tests ", ($network_tests =~ /y/i) ? 1 : 0, "\n";
    close OUT;
}

## HELPERS

sub Candidate {
    my $dir = shift;
    return unless -d $dir;

    my $version_file;
    for (
         "$dir/inc32/openssl/opensslv.h", # old win32 builds
         "$dir/crypto/opensslv.h", # cygwin32 builds
         "$dir/include/openssl/opensslv.h",
         "$dir/../include/openssl/opensslv.h", # Solaris
         "$dir/include/opensslv.h",
         "$dir/include/crypto.h"
    ) {
        if (-e $_) {
            $version_file = $_;
            last;
        }
    }
    return unless defined $version_file;
    my $fingerprint = join(':', (stat $version_file)[0,1]);

    my $inc_dir = dirname($version_file);
    return unless -e "$inc_dir/ssl.h";

    my $prefix;
    if ( 'openssl' eq lc basename $inc_dir ) {
        $prefix = 'openssl/';
        $inc_dir = dirname $inc_dir;
    }
    else {
        $prefix = '';
    }

    open(VERSION_FILE, $version_file) or return;
    my $version;
    my $type;

    # Changed version number matching because
    # #define SSLEAY_VERSION_NUMBER   OPENSSL_VERSION_NUMBER
    # in crypto/crypto.h in at least some early versions of OpenSSL
    while (<VERSION_FILE>) {
        next unless my ($vstr) =
        /^#define\s+(?:OPENSSL|SSLEAY)_VERSION_NUMBER\s+0x([0-9A-Fa-f]+)/;

        ($type, $version) = parse_openssl_version( $vstr );
        die "Failed to parse OpenSSL version '$vstr'" unless defined $version;
        last;
    }
    close(VERSION_FILE);
    die "Failed to identify OpenSSL version" unless defined $version;

    # Silly test to look for the library files
    my $found_lib;
    my $libd;
    my $subdir = $opt_static ? 'out32' : 'out32dll';
    if (-d "$dir/$subdir") {
        $libd = [$subdir];
    }
    elsif ($^O eq 'MSWin32' and $Config{cc} eq 'gcc') {
        $libd = ['lib/MinGW'];
    }
    else {
        # second attempt is for Solaris, like the include directory, the
        # library directory may be in a sibling directory
        $libd = ['lib', '../lib'];
    }

    SCAN:
    for my $d (@$libd) {
        my $lib_dir = "$dir/$d";
        if (opendir(LIBDIR, $lib_dir)) {
            while (defined($_ = readdir(LIBDIR))) {
                if (/\A(?:lib(?:crypto|eay32|ssl)|ssleay32)/) {
                    $found_lib = $lib_dir;
                    last SCAN;
                }
            }
            closedir(LIBDIR);
        }
    }

    if (!$found_lib) {
        my @tried = join( ',' => map {"$dir/$_"} @$libd);
        print "Did not locate expected SSL library files in @tried\n";
    }

    return {
        dir    => $dir,
        inc    => $inc_dir,
        lib    => $found_lib,
        ver    => $version,
        type   => $type,
        prefix => $prefix,
        check  => $fingerprint,
    };
}

sub parse_openssl_version {
    my ($vstr) = @_;

    # 0.9.6 or later : MNNFFPPS : major, minor, fix, patch, status (0-f)
    # 0.9.5a to before 0.9.6 : high bit of PP is set in MNNFFPPS
    # 0.9.3-dev to before 0.9.5a : MNNFFRBB : major, minor, fix, status (0-1), beta
    # prior to 0.9.3-dev, MNFP

    if ( $vstr =~ /^(?:[1-9]|0090[6-9])/ ) {
        return ( OpenSSL => parse_openssl_v096_or_later( $vstr ) );
    }
    elsif ( $vstr =~ /^00905[8-9]/ ) {
        return ( OpenSSL => parse_openssl_v095a_096( $vstr ) );
    }
    elsif ( $vstr =~ /^0090[3-5][01]/ ) {
        return ( OpenSSL => parse_openssl_v093dev_095a( $vstr ) );
    }
    else {
        return ( SSLeay => parse_openssl_pre_v093dev( $vstr ) );
    }
    return;
}

sub parse_openssl_v096_or_later {
    my ($vstr) = @_;

    # 0.9.6 or later : MNNFFPPS : major, minor, fix, patch, status (0-f)
    return unless my @v = $vstr =~ /^(.)(..)(..)(..)(.)$/;

    my ($major, $minor, $fix, $patch, $status) = map hex, @v;

    $patch = $patch ? chr( ord('a') + $patch - 1 ) : '';

    if    ( $status ==  0 ) { $status = '-dev' }
    elsif ( $status == 15 ) { $status = '' }
    else                    { $status = "-beta$status" }
    return sprintf( '%u.%u.%u%s%s', $major, $minor, $fix, $patch, $status);
}

sub parse_openssl_v095a_096 {
    my ($vstr) = @_;

    # 0.9.5a to before 0.9.6 : high bit of PP is set in MNNFFPPS

    return parse_openssl_v096_or_later(
        sprintf '%08x', hex($vstr) & 0xfffff7ff
    );
}

sub parse_openssl_v093dev_095a {
    my ($vstr) = @_;

    # 0.9.3-dev to before 0.9.5a : MNNFFRBB : major, minor, fix, status (0-1), beta

    return unless my @v = $vstr =~ /^(0)(09)(0[3-5])([0-1])(..)$/;

    my ($major, $minor, $fix, $status, $patch) = map hex, @v;

    $patch = $patch ? chr( ord('a') + $patch - 1 ) : '';
    $status = $status ? '' : '-dev';
    return sprintf( '%u.%u.%u%s%s', $major, $minor, $fix, $patch, $status );
}

sub parse_openssl_pre_v093dev {
    my ($vstr) = @_;

    # prior to 0.9.3-dev, MNFP

    return unless my @v = $vstr =~ /^(0)(9)([12])([0-9])$/;

    my ($major, $minor, $fix, $patch) = map hex, @v;

    $patch = $patch ? chr( ord('a') + $patch - 1 ) : '';
    return sprintf( '%u.%u.%u%s', $major, $minor, $fix, $patch );
}


